tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::registry::SpanRef;
+use tracing_subscriber::Layer;
+use std::sync::Arc;
+
+pub struct FlameLayer<S, W>
+where
+    S: Subscriber,
+    W: Write + 'static,
+{
+    out: Arc<Mutex<W>>,
+    last_event: Mutex<Instant>,
+    _inner: PhantomData<S>,
+}
+
+pub struct FlameGuard<W>
+where
+    W: Write + 'static,
+{
+    out: Arc<Mutex<W>>,
+}
+
+impl<S, W> FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
+{
+    pub fn new(writer: W) -> Self {
+        Self {
+            out: Arc::new(Mutex::new(writer)),
+            last_event: Mutex::new(Instant::now()),
+            _inner: PhantomData,
+        }
+    }
+
+    pub fn flush_on_drop(&self) -> FlameGuard<W> {
+        FlameGuard {
+            out: self.out.clone(),
+        }
+    }
+}
+
+impl<W> Drop for FlameGuard<W>
+where
+    W: Write + 'static,
+{
+    fn drop(&mut self) {
+        dbg!(());
+        match self.out.lock().unwrap().flush() {
+            Ok(_) => (),
+            Err(e) => eprintln!("Error: {}", e),
+        }
+    }
+}
+
+impl<S> FlameLayer<S, BufWriter<File>>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+{
+    pub fn write_to_file(path: impl AsRef<Path>) -> Result<Self, Error> {
+        let path = path.as_ref();
+        let file = File::create(path).map_err(|source| Error::IO {
+            path: path.into(),
+            source,
+        })?;
+        let writer = BufWriter::new(file);
+        Ok(Self::new(writer))
+    }
+}
+
+impl<S, W> Layer<S> for FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
+{
+    fn new_span(&self, _: &Attributes, id: &Id, ctx: Context<S>) {
+        let samples = self.time_since_last_event();
+
+        let parent = ctx.span(id).expect("expected: span id exists in registry").parent();
+
+        let mut stack = String::new();
+
+        stack.push_str("tracing");
+
+        if let Some(parent) = parent {
+            for parent in parent.from_root() {
+                stack.push_str("; ");
+                write(&mut stack, parent).expect("expected: write to String never fails");
+            }
+        }
+
+        write!(&mut stack, " {}", samples.as_nanos())
+            .expect("expected: write to String never fails");
+        writeln!(
+            *self.out.lock().unwrap(),
+            "{}",
+            stack
+        )
+        .expect("expected: write to String never fails");
+    }
+
+    #[allow(clippy::needless_return)]
+    fn on_close(&self, id: Id, ctx: Context<S>) {
+        let panicking = std::thread::panicking();
+        macro_rules! expect {
+            ($e:expr, $msg:literal) => {
+                if panicking {
+                    return;
+                } else {
+                    $e.expect($msg)
+                }
+            };
+        }
+
+        let samples = self.time_since_last_event();
```

Comment:
this method can panic:
https://github.com/tokio-rs/tracing/blob/1639b8449e326f87a8bc9431e314413d4c2154e2/tracing-flame/src/lib.rs#L250

I think (unfortunately), in order to avoid double panics, you're going to need to have `time_since_last_event` return an `Option` or `Result`, and have the code calling it expect that the mutex isn't poisoned. Or, the function could take a `MutexGuard` instead of a `self` reference, and then callers can decide how to handle a poisoned mutex.